sys/evtimer/evtimer.c

```diff
@@ -0,0 +1,185 @@
+    }
+}
+
+static void _set_timer(xtimer_t *timer, uint32_t offset)
+{
+    uint64_t offset_in_us = (uint64_t)offset * 1000;
+    DEBUG("evtimer: now=%"PRIu32" setting xtimer to %"PRIu32":%"PRIu32"\n",  xtimer_now(),
+            (uint32_t)(offset_in_us>>32), (uint32_t)(offset_in_us));
+    _xtimer_set64(timer, offset_in_us, offset_in_us>>32);
+}
+
+static void _update_timer(evtimer_t *evtimer)
+{
+    if (evtimer->events) {
+        evtimer_event_t *event = evtimer->events;
+        _set_timer(&evtimer->timer, event->offset);
+    }
+    else {
+        xtimer_remove(&evtimer->timer);
+    }
+}
+
+static uint32_t _get_offset(xtimer_t *timer)
+{
+    uint64_t now = xtimer_now64();
+    uint64_t target = ((uint64_t)timer->long_target)<<32 | timer->target;
+
+    if (target <= now) {
+        return 0;
+    }
+    else {
+        target -= now;
+        /* add half of 125 so integer division rounds to nearest */
+        return div_u64_by_125((target >> 3) + 62);
+    }
+}
+
+static void _update_head_offset(evtimer_t *evtimer)
+{
+    if (evtimer->events) {
+        evtimer_event_t *event = evtimer->events;
+        event->offset = _get_offset(&evtimer->timer);
+        DEBUG("evtimer: _update_head_offset(): new head offset %"PRIu32"\n", event->offset);
+    }
+}
+
+void evtimer_add(evtimer_t *evtimer, evtimer_event_t *event)
+{
+    unsigned state = irq_disable();
+    DEBUG("evtimer_add(): adding event with offset %"PRIu32"\n", event->offset);
+
+    _update_head_offset(evtimer);
+    _add_event_to_list((evtimer_event_t*)&evtimer->events, event);
+
+    if (evtimer->events == event) {
+        _set_timer(&evtimer->timer, event->offset);
+    }
+    irq_restore(state);
+}
+
+void evtimer_del(evtimer_t *evtimer, evtimer_event_t *event)
+{
+    unsigned state = irq_disable();
+    DEBUG("evtimer_del(): removing event with offset %"PRIu32"\n", event->offset);
+    _update_head_offset(evtimer);
+    _del_event_from_list((evtimer_event_t*)&evtimer->events, event);
+    _update_timer(evtimer);
+    irq_restore(state);
+}
+
+static evtimer_event_t *_get_next(evtimer_t *evtimer)
+{
+    evtimer_event_t *event = evtimer->events;
+    if (event && (event->offset == 0)) {
+        evtimer->events = event->next;
+        return event;
+    }
+    else {
+        return NULL;
+    }
+}
+
+void evtimer_msg_handler(void *arg)
+{
+    DEBUG("evtimer_msg_handler()\n");
+
+    evtimer_t *evtimer = (evtimer_t*) arg;
+
+    evtimer_event_t *event = evtimer->events;
+    event->offset = 0;
+
+    evtimer_msg_event_t *mevent;
+    while ((mevent = (evtimer_msg_event_t *)_get_next(evtimer))) {
+        msg_send_int(&mevent->msg, mevent->msg.sender_pid);
+    }
+
+    _update_timer(evtimer);
+
+    if (!irq_is_in()) {
+        thread_yield_higher();
+    }
+}
+
+void evtimer_init(evtimer_t *evtimer, void(*handler)(void*))
+{
+    evtimer->timer.callback = handler;
+    evtimer->timer.arg = (void*) evtimer;
+    evtimer->events = NULL;
+}
+
+#if ENABLE_DEBUG == 1
```

Comment:
This requires any external caller to edit this file internally (which also causes printing of the messages above). The compile should optimize it away if it isn't called. Alternatively, put this function in another file (if you don't trust the compiler optimization).